### PR TITLE
OCPBUGS-18093:Allow users to specify Gateway Subnet

### PR DESCRIPTION
**Pull request by bpickard22**
We currently reserve a range of addresses to configure host to service traffic internally in ovn. We need to allow users to specify this range to avoid conflicting with addresses they use in their local infra

### CODE REVIEW

**Comment by abhat** [comment]
LGTM

**Comment by abhat** [comment]
need to do a codegen since the verify job is failing?

File: operator/v1/types_network.go
Comment (tssurya) [inline operator/v1/types_network.go:468]: nit: These are kubebuilder formats, please follow convention and put the comment above the field.

File: operator/v1/0000_70_cluster-network-operator_01.crd.yaml
Comment (tssurya) [inline operator/v1/0000_70_cluster-network-operator_01.crd.yaml:287]: did you manually edit this or did this get auto-generated? `gatewaySubnet` is what I'd expect...

File: operator/v1/types_network.go
Comment (tssurya) [reply operator/v1/types_network.go:468]: Example start the comment with `gatewaySubnet contains..` blah blah...

File: operator/v1/types_network.go
Comment (tssurya) [inline operator/v1/types_network.go:462]: ah I see why now.. 
```suggestion
	GatewaySubnet  string         `json:"gatewaySubnet,omitempty"`
```
please pay attention to semantics, notice how other fields are being done..

Comment (tssurya) [inline operator/v1/types_network.go:462]: Suggestions:
Should we put these new options inside the `GatewayConfig`?
And shouldn't we have 2 subnet options one for v4 and another for v6 ? How do you plan to have both of them in this single field?

169.254.169.0/16 isn't exactly a Gateway Subnet, if we move this into `GatewayConfig` then we could call them `V4InternalMasqueradeSubnet` and `V6InternalMasqueradeSubnet`..

**Review by tssurya — CHANGES_REQUESTED** [review]
left a few comments, but I know we are waiting for the bits from the generated code..

File: operator/v1/0000_70_cluster-network-operator_01.crd.yaml
Comment (bpickard22) [reply operator/v1/0000_70_cluster-network-operator_01.crd.yaml:287]: I did that manually

File: operator/v1/types_network.go
Comment (bpickard22) [reply operator/v1/types_network.go:462]: will correct

Comment (bpickard22) [reply operator/v1/types_network.go:462]: yes I think this makes more sense, will use this approach

File: operator/v1/types_network.go
Comment (tssurya) [inline operator/v1/types_network.go:489]: We also use the `169.254.169.5` btw for the hairpin service traffic. Since its a subnet, its enough to say 
`Default is 169.254.169.0/16` or whatever is the default "subnet".
Doesn't make sense to quote the IPs within the subnet as defaults right?

File: operator/v1/types_network.go
Comment (tssurya) [inline operator/v1/types_network.go:496]: ditto

File: operator/v1/types_network.go
Comment (tssurya) [reply operator/v1/types_network.go]: cc @trozet if he is ok with this.

**Review by tssurya — CHANGES_REQUESTED** [review]
Almost there.
BTW this can't merge without the OVNK PR right? Probably best to put a hold until then?

File: operator/v1/types_network.go
Comment (bpickard22) [reply operator/v1/types_network.go:489]: > Almost there. BTW this can't merge without the OVNK PR right? Probably best to put a hold until then?

right, just wanted this up for reivew

Comment (bpickard22) [reply operator/v1/types_network.go:489]: @tssurya should we just call it a range then. If the user needs to use those addresses for their infra, are we going to shift the masq addresses to a different subnet or just silde them down in the same subnet?

Comment (trozet) [reply operator/v1/types_network.go:489]: I think it should be a subnet. The defaults right now are:
	V4MasqueradeSubnet              = "169.254.169.0/29"
	V6MasqueradeSubnet              = "fd69::/125"

We can specify that the minimum cidr needs to be at least /29 for v4 and at least /125 for v6

**Comment by trozet** [comment]
I thought we also need the ability to specify the join subnet?

**Comment by trozet** [comment]
> I thought we also need the ability to specify the join subnet?

@bpickard22 pointed out it already exists via: https://github.com/openshift/api/pull/1207

File: operator/v1/types_network.go
Comment (bpickard22) [reply operator/v1/types_network.go]: @trozet is there a reason that we need to specify that min cidr?

File: operator/v1/types_network.go
Comment (deads2k) [inline operator/v1/types_network.go:510]: Under what conditions would a cluster-admin want to change this and how would they know if they were choosing a "good" value?  How would they know if their choice was bad?

File: operator/v1/types_network.go
Comment (deads2k) [inline operator/v1/types_network.go:514]: I think there's a regex that can validate this.  I bet @Miciah knows for this and for the ipv6 one.

File: operator/v1/types_network.go
Comment (bpickard22) [reply operator/v1/types_network.go:510]: A cluster admin would want to change this if they want to addresses that we currently use here in their local infrastructure. I think "good" may be subjective, but any address range  that neither we nor the cloud provider reserve should be good

File: operator/v1/types_network.go
Comment (bpickard22) [reply operator/v1/types_network.go:514]: we dont currently use regex to validate the other configurable subnet in ovn-k (the internal join subnet), and any invalid value passed in would be caught by the cno or in ovnk, do you think we need that here?

File: operator/v1/types_network.go
Comment (bpickard22) [reply operator/v1/types_network.go:511]: the host networked pod

File: openapi/openapi.json
Comment (flavio-fernandes) [inline openapi/openapi.json:25121]: NIT: add a space after period "... gateway bridge interface. The values can be changed after ..."

File: operator/v1/0000_70_cluster-network-operator_01.crd.yaml
Comment (flavio-fernandes) [inline operator/v1/0000_70_cluster-network-operator_01.crd.yaml:227]: ditto on the space after '.'

File: operator/v1/0000_70_cluster-network-operator_01.crd.yaml
Comment (flavio-fernandes) [inline operator/v1/0000_70_cluster-network-operator_01.crd.yaml:230]: ditto on the space after '.'

File: operator/v1/types_network.go
Comment (trozet) [reply operator/v1/types_network.go:510]: I think we can just say: The subnet chosen should not overlap with other networks specified for OVN-Kubernetes as well as other networks used on the host. Additionally the subnet must be large enough to accommodate 6 IPs (maximum prefix length /29).

Comment (bpickard22) [reply operator/v1/types_network.go:510]: Updated the description with this addition

File: operator/v1/types_network.go
Comment (bpickard22) [reply operator/v1/types_network.go:516]: added regex taken from https://github.com/openshift/api/blob/3e3f07aadec46cad1812b122a51ce83559934054/operator/v1/types_ingress.go#L378-L386

Comment (bpickard22) [reply operator/v1/types_network.go:516]: split into v4 and v6 respectively

File: operator/v1/types_network.go
Comment (JoelSpeed) [inline operator/v1/types_network.go:510]: Godoc should start with the json tag name, not the Go type name
```suggestion
	// v4InternalMasqueradeSubnet contains the v4 masquerade addresses used internally by
```

Comment (JoelSpeed) [inline operator/v1/types_network.go:510]: Do you expect in the future to add any other config to this, that might be split into IPv4 and IPv6 versions? Wondering if creating some structure now will make this easier to consume in the future?
```
gatewayConfig:
  ipForwarding: Restricted
  ipv4:
    internalMasqueradeSubnet: ...
  ipv6:
    internalMasqueradeSubnet: ...
```

File: operator/v1/types_network.go
Comment (JoelSpeed) [reply operator/v1/types_network.go:511]: We should update the description, perhaps `Each host in the cluster is configured...` covers this?

File: operator/v1/types_network.go
Comment (JoelSpeed) [inline operator/v1/types_network.go:515]: Does the regex account for this minimum?

File: operator/v1/types_network.go
Comment (JoelSpeed) [inline operator/v1/types_network.go:516]: I think to fit in with standard prose
```suggestion
	// When omitted, this means no opinion and the platform is left to choose a reasonable default which is subject to change over time. Currently the default subnet is 169.254.169.0/29.
```

Comment (JoelSpeed) [inline operator/v1/types_network.go:516]: The format is not explained within the godoc, it must be spelt out otherwise users won't know the validation. I guess for this it's sufficient to say that the value must be in CIDR format?

File: operator/v1/types_network.go
Comment (JoelSpeed) [inline operator/v1/types_network.go:517]: Is `[0-9][0-9]` correct as an option? Would you ever have a leading zero in a two digit octet?

Comment (JoelSpeed) [inline operator/v1/types_network.go:517]: You don't want `/30`, `/31` or `/32` right? So perhaps we drop `3[0-2]`?

Comment (JoelSpeed) [inline operator/v1/types_network.go:517]: Please add some integration tests base on [this documentation](https://github.com/openshift/api/blob/master/tests/README.md) to test the API validation here

**Review by JoelSpeed — COMMENTED** [review]
I've only really gone into the detail on the ipv4 version of this, I suspect most of my comments apply to the v6 version too

File: operator/v1/types_network.go
Comment (bpickard22) [reply operator/v1/types_network.go:517]: it would be a valid class a address I believe so I think thats okay

File: operator/v1/types_network.go
Comment (bpickard22) [reply operator/v1/types_network.go:515]: correcting to account for this now

File: operator/v1/types_network.go
Comment (JoelSpeed) [reply operator/v1/types_network.go]: I wonder if we should go ahead and do it anyway, just in case, IMO it's cleaner and more future proof, WDYT?

File: operator/v1/types_network.go
Comment (JoelSpeed) [reply operator/v1/types_network.go]: Integration tests for the validations are outstanding as well, please add

Comment (bpickard22) [reply operator/v1/types_network.go]: Okay I will change to not allow leading zeros

File: operator/v1/types_network.go
Comment (bpickard22) [reply operator/v1/types_network.go]: Agree on the future proof aspect. I dont currently see a scenario in which we would be adding anything to the gateway config, but that could change, so I can make this change

File: operator/v1/types_network.go
Comment (JoelSpeed) [inline operator/v1/types_network.go:518]: This regex is not very, human readable, no regex is. When an error occurs, the regex is printed to the end user, which, doesn't necessarily help them to understand why their input value was wrong.

As an alternative, I think we can break this down into CEL validations that help the user, take a look at [this commit](https://github.com/openshift/api/commit/49db644cc07486559bbfe3148868ff27ace1be9d) that I put together with some CEL validations that check for a valid IPv4 address and some tests to show the kind of thing we want there

File: operator/v1/types_network.go
Comment (JoelSpeed) [inline operator/v1/types_network.go:532]: IIUC the maximum length of an IPv6 address is 48 characters, then plus four to add the `/xxx` subnet, this string should have a maximum length of 52 characters right?
```suggestion
	// +kubebuilder:validation:MaxLength:=52
	// +kubebuilder:validation:Pattern=`(^s*((([0-9A-Fa-f]{1,4}:){7}([0-9A-Fa-f]{1,4}|:))|(([0-9A-Fa-f]{1,4}:){6}(:[0-9A-Fa-f]{1,4}|((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3})|:))|(([0-9A-Fa-f]{1,4}:){5}(((:[0-9A-Fa-f]{1,4}){1,2})|:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3})|:))|(([0-9A-Fa-f]{1,4}:){4}(((:[0-9A-Fa-f]{1,4}){1,3})|((:[0-9A-Fa-f]{1,4})?:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){3}(((:[0-9A-Fa-f]{1,4}){1,4})|((:[0-9A-Fa-f]{1,4}){0,2}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){2}(((:[0-9A-Fa-f]{1,4}){1,5})|((:[0-9A-Fa-f]{1,4}){0,3}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){1}(((:[0-9A-Fa-f]{1,4}){1,6})|((:[0-9A-Fa-f]{1,4}){0,4}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:))|(:(((:[0-9A-Fa-f]{1,4}){1,7})|((:[0-9A-Fa-f]{1,4}){0,5}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:)))(%.+)?s*(/(12[0-5]|1[0-1][0-9]|[1-9][0-9]|[0-9]))$)`
```

File: operator/v1/stable.network.testsuite.yaml
Comment (JoelSpeed) [inline operator/v1/stable.network.testsuite.yaml:196]: Nit, all files should end with a new line character, git can produce weird diffs otherwise

File: operator/v1/types_network.go
Comment (JoelSpeed) [inline operator/v1/types_network.go:513]: Shouldn't be a pointer unless it contains required fields, also doesn't need omitempty
```suggestion
	IPv4 IPv4GatewayConfig `json:"ipv4"`
```

File: operator/v1/types_network.go
Comment (JoelSpeed) [inline operator/v1/types_network.go:510]: Typo
```suggestion
	// ipv4 allows users to configure IP settings for IPv4 connections. When omitted, this means no opinion and the default
```

File: operator/v1/types_network.go
Comment (JoelSpeed) [inline operator/v1/types_network.go:517]:
```suggestion
	// ipv6 allows users to configure IP settings for IPv6 connections. When omitted, this means no opinion and the default
	// configuration is used. Check individual members fields within ipv6 for details of default values.
	// +optional
	IPv6 IPv6GatewayConfig `json:"ipv6"`
```

File: operator/v1/types_network.go
Comment (JoelSpeed) [inline operator/v1/types_network.go:520]: Struct should have a godoc

File: operator/v1/types_network.go
Comment (JoelSpeed) [inline operator/v1/types_network.go:521]: We don't need to duplicate that `v4` anymore, and we still don't mention that it's in CIDR format
```suggestion
	// internalMasqueradeSubnet contains the masquerade addresses in IPv4 CIDR format, used internally by
```

File: operator/v1/types_network.go
Comment (JoelSpeed) [inline operator/v1/types_network.go:538]: These comments can now be removed since we got rid of the regex right?

File: operator/v1/types_network.go
Comment (JoelSpeed) [inline operator/v1/types_network.go:539]: No need for v4 prefix or omitempty
```suggestion
	InternalMasqueradeSubnet string `json:"InternalMasqueradeSubnet"`
```

File: operator/v1/types_network.go
Comment (JoelSpeed) [inline operator/v1/types_network.go:542]: Should have godoc on the struct

File: operator/v1/types_network.go
Comment (JoelSpeed) [inline operator/v1/types_network.go:543]:
```suggestion
	// internalMasqueradeSubnet contains the masquerade addresses in IPv6 CIDR format, used internally by
```

File: operator/v1/types_network.go
Comment (JoelSpeed) [inline operator/v1/types_network.go:549]: I think here we need to mention dual addresses are forbidden
```suggestion
	// Note that IPv6 dual addresses are not permitted.
	// When omitted, this means no opinion and the platform is left to choose a reasonable default which is subject to change over time.
```

We also need a new validation for this
```
// +kubebuilder:validation:XValidation:rule="!self.contains('.')",message="IPv6 dual addresses are not permitted, value should not contain `.` characters"
```

File: operator/v1/types_network.go
Comment (JoelSpeed) [inline operator/v1/types_network.go:554]: This rule currently allows for dual addresses
```suggestion
	// +kubebuilder:validation:XValidation:rule="self.split('/')[0].split(':').size() <= 8",message="a valid IPv6 address must contain at most 8 segments"
```

File: operator/v1/types_network.go
Comment (JoelSpeed) [inline operator/v1/types_network.go:567]: These comments are no longer required

**Comment by abhat** [comment]
@JoelSpeed thank you for the thorough review. Is this PR good to go from your perspective? Looks like @bpickard22 has addressed most of the comments and has the integration tests in place as well.

cc @trozet

**Comment by JoelSpeed** [comment]
Yep, I'm good to label this, is the implementation PR ready to merge once this is done? Are we planning to merge during the bug window that's currently in place for 4.14?

**Comment by trozet** [comment]
@JoelSpeed the implementation is ready. We can merge during the bug window (either way we will need a jira bug) as it is not a risky change to CNO and the OVNK bits have already landed downstream.

**Comment by JoelSpeed** [comment]
Ack, @bpickard22 Can you link this and the implementation up to a bug so they are tied together, than we can merge if everyone is happy?

@trozet You happy to add LGTM? I'll add approve once that and bug link are done

**Review by trozet — APPROVED** [review]
/lgtm

**Comment by bpickard22** [comment]
/retitle OCPBUGS-18093:Allow users to specify Gateway Subnet

**Pull request merged**
